### PR TITLE
WIP: #190 JSON export

### PR DESCRIPTION
**Pull request by GneyHabub**
Now there are 2 buttons over each _Content_ field of the Messages table. 

image

If you hover over any of them there will be a popup hint which tells the user what this button does:

image
image

The first button will copy the Content to the clipboard and the second will trigger a file download (`.json` if the content is an object or `JSON.parse` doesn't throw an exception; `.txt` otherwise.)

### CODE REVIEW

**Comment by workshur** [comment]
1. Please add description 
2. Add screenshot

File: kafka-ui-react-app/src/components/App.scss [hunk 1]
```diff
@@ -36,3 +36,26 @@ $navbar-width: 250px;
...
+    pointer-events: none;
+    transition: .2s opacity;
+    transition-delay: 0s;
+  }
+
+  &:hover::after {
+    opacity: 1;
+    transition-delay: .5s;
+   }
+}
```
[the remark this answers is not in the document]
Comment (workshur) [reply kafka-ui-react-app/src/components/App.scss:61]: It is important to reduce custom stuff. Let's use native title

**Review by workshur — CHANGES_REQUESTED** [review]
1. We need to move buttons to JSONViewer component.
2. Use standard bulma buttons

**Comment by GneyHabub** [comment]
I have no idea how to test the file download yet. Still working on it.

**Comment by GneyHabub** [comment]
Okay, after some research I've got an impression that the file-downloading part of the functionality is pretty much untestable. I didn't find a way to trigger the file download inside of the testing environment. 
It is possible to dig into the specific implementation and test it, but this seems wrong to me. In this way, we treat the component as a white box, exposing all of its guts, tying our tests to the specific implementation. Looks like a violation of a bunch of principles and good practises. 
@workshur @whotake what do you think?

**Review by workshur — APPROVED** [review]

File: kafka-ui-react-app/src/components/Topics/Details/Messages/MessageItem.tsx [hunk 2]
```diff
@@ -21,7 +21,11 @@ const MessageItem: React.FC<MessageItemProp> = ({
     <td style={{ width: 150 }}>{offset}</td>
     <td style={{ width: 100 }}>{partition}</td>
     <td style={{ wordBreak: 'break-word' }}>
-      {content && <JSONViewer data={content as { [key: string]: string }} />}
+      {content && (
+        <div>
+          <JSONViewer data={content as { [key: string]: string }} />
```
Comment (whotake) [inline kafka-ui-react-app/src/components/Topics/Details/Messages/MessageItem.tsx:26]: tip: `Record<string, string>`

File: kafka-ui-react-app/src/components/common/JSONViewer/JSONViewer.tsx [hunk 3]
```diff
@@ -1,15 +1,60 @@
...
+  const { copyToClipboard, saveFile } = useDataSaver();
+  const copyButtonHandler = () => {
+    copyToClipboard(JSON.stringify(data));
+  };
+  const buttonClasses = 'button is-link is-outlined is-small is-centered';
+  return (
+    <div>
+      <div
+        className="field has-addons"
+        style={{
```
Comment (whotake) [inline kafka-ui-react-app/src/components/common/JSONViewer/JSONViewer.tsx:23]: You can use Bulma helpers https://bulma.io/documentation/helpers/flexbox-helpers/#justify-content

File: kafka-ui-react-app/src/components/common/JSONViewer/JSONViewer.tsx [hunk 4]
```diff
@@ -1,15 +1,60 @@
...
+          justifyContent: 'flex-end',
+        }}
+      >
+        <DynamicButton
+          callback={copyButtonHandler}
+          classes={buttonClasses}
+          title="Copy the message to the clipboard"
+          style={{ marginRight: '5px' }}
```
Comment (whotake) [inline kafka-ui-react-app/src/components/common/JSONViewer/JSONViewer.tsx:31]: Same here

**Comment by whotake** [comment]
@GneyHabub Does it throw any errors when are you trying to test it?

I think we can split view and download logic and test them separately:

1. Test view that it calls `saveFile` with proper arguments
2. Test `saveFile` itself. And there a few options on how to do it:
2.1 Split function into modules and test important logic. We care about file name and content more than pretty much standard downloading function.
2.2 Mock `document` and check `setAttribute` to content
2.3 Combine two methods above

File: kafka-ui-react-app/src/components/Topics/Topic/Details/Messages/MessageItem.tsx [hunk 5]
```diff
@@ -21,7 +21,11 @@ const MessageItem: React.FC<MessageItemProp> = ({
     <td style={{ width: 150 }}>{offset}</td>
     <td style={{ width: 100 }}>{partition}</td>
     <td style={{ wordBreak: 'break-word' }}>
-      {content && <JSONViewer data={content as { [key: string]: string }} />}
+      {content && (
+        <div>
+          <JSONViewer data={content as Record<string, string>} />
+        </div>
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Topics/Topic/Details/Messages/MessageItem.tsx:27]: What is this div for?

Comment (GneyHabub) [reply kafka-ui-react-app/src/components/Topics/Topic/Details/Messages/MessageItem.tsx]: Yep, got rid of it

**Pull request merged**